File: 0986-interval-list-intersections/0986-interval-list-intersections.go
func intervalIntersection(firstList [][]int, secondList [][]int) [][]int {
    ans := [][]int{}
    f, s := 0, 0
    
    for f < len(firstList) && s < len(secondList) {
        start := max(firstList[f][0], secondList[s][0])
        end := min(firstList[f][1], secondList[s][1])
        
        if start <= end {
            ans = append(ans, []int{start, end})
        }
        
        if firstList[f][1] < secondList[s][1] {
            f++
        } else {
            s++
        }
    }
    
    return ans
}

func min(i, j int) int {
    if i < j {
        return i
    }
    
    return j
}

func max(i, j int) int {
    if i > j {
        return i
    }
    
    return j
}

File: 0986-interval-list-intersections/0986-interval-list-intersections.php
class Solution {

    /**
     * @param Integer[][] $firstList
     * @param Integer[][] $secondList
     * @return Integer[][]
     */
    function intervalIntersection($firstList, $secondList) {
        $ans = [];
        $first = $second = 0;
        
        while($first < count($firstList) && $second < count($secondList)) {
            $start = max($firstList[$first][0], $secondList[$second][0]);
            $end = min($firstList[$first][1], $secondList[$second][1]);
            
            if($start <= $end) {
                $ans[] = [$start, $end];
            }
            
            $firstList[$first][1] < $secondList[$second][1] ? $first++ : $second++;
        }
        
        return $ans;
    }
}